[DOCS] Update remote cluster docs (#77043)
elastic/elasticsearch: docs/reference/modules/cluster/remote-clusters-connect.asciidoc

[[remote-clusters-connect]]
=== Connect to remote clusters

> Your local cluster uses the <<modules-network,transport layer>> to establish

Reviewer: ```suggestion
Your local cluster uses the <<modules-network,transport interface>> to establish
```

The linked networking doc uses the phrase "transport interface", which I like better for clarity than either "transport layer" or "transport level". I think we should strive for consistency here.

communication with remote clusters. The coordinating nodes in the local cluster
establish <<long-lived-connections,long-lived>> TCP connections with specific
nodes in the remote cluster. {es} requires these connections to remain open,
even if the connections are idle for an extended period.

NOTE: You must have the `manage` cluster privilege to connect remote clusters.

To configure a remote cluster from Stack Management in {kib}:

. Select *Remote Clusters* from the side navigation.
. Specify the {es} endpoint URL, or the IP address or host name of the remote
cluster followed by the transport port (defaults to `9300`). For example,
`cluster.es.eastus2.staging.azure.foundit.no:9400` or `192.168.1.1:9300`.

Alternatively, use the <<cluster-update-settings,cluster update settings API>>

> to add a remote cluster. You can also use this API to
> <<configure-remote-clusters-dynamic,apply and update>> settings across the
> entire remote cluster. To configure settings on individual nodes in a remote
> cluster, define
> <<configure-remote-clusters-static,static settings>> in `elasticsearch.yml`.

Reviewer: This still reads funny to me. I think what we want to say is:

```suggestion
to add a remote cluster. You use this API to dynamically
<<configure-remote-clusters-dynamic,configure remote clusters>> for every node 
in the local cluster. To configure remote clusters for an individual node of the local cluster,
define <<configure-remote-clusters-static,static settings>> in the node's `elasticsearch.yml`.
```

Author: I'll change this to read:

> Alternatively, use the <<cluster-update-settings,cluster update settings API>>
> to add a remote cluster. You use this API to dynamically
> <<configure-remote-clusters-dynamic,configure remote clusters>> for _every_ node
> in the local cluster. To configure remote clusters for individual nodes in the
> local cluster, define <<configure-remote-clusters-static,static settings>> in `elasticsearch.yml` for each node.

> <<configure-remote-clusters-static,static settings>> in `elasticsearch.yml`.

Author: As always, I appreciate your in-depth explanations 👍 I'll change that last sentence to read:
>To configure settings on individual nodes in a remote cluster, define static settings in `elasticsearch.yml`.

The following request adds a remote cluster with an alias of `cluster_one`. This
_cluster alias_ represents the connection to the remote cluster and is used 
subsequently to distinguish between local and remote indices. When adding a

> remote cluster in {kib}, the cluster alias serves as a unique identifier for the

Author: I'm actually going to remove this sentence and combine "unique identifier" into the previous sentence:
>The following request adds a remote cluster with an alias of `cluster_one`. This _cluster alias_ is a unique identifier that represents the connection to the remote cluster and is used subsequently to distinguish between local and remote indices.

configure remote settings on every node in the cluster. The following request
adds three remote clusters: `cluster_one`, `cluster_two`, and `cluster_three`.

The `seeds` parameter specifies the hostname and
<<transport-settings,transport port>> (default `9300`) of a seed node in the 
remote cluster.

The `mode` parameter determines the configured connection mode, which defaults
to <<sniff-mode,`sniff`>>. Because `cluster_one` doesn't specify a `mode`, it
uses the default. Both `cluster_two` and `cluster_three` explicitly use
different modes.

[source,console]
----
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "cluster_one": {
          "seeds": [
            "127.0.0.1:9300"
          ]
        },
        "cluster_two": {
          "mode": "sniff",
          "seeds": [
            "127.0.0.1:9301"
          ],
          "transport.compress": true,
          "skip_unavailable": true
        },
        "cluster_three": {
          "mode": "proxy",
          "proxy_address": "127.0.0.1:9302"
        }
      }
    }
  }
}
----
// TEST[setup:host]
// TEST[s/127.0.0.1:9300/\${transport_host}/]

> You can dynamically update other cluster settings for a cluster, such as
> <<transport-settings,transport settings>>. The following request updates the

Reviewer: ```suggestion
You can dynamically update settings for a remote cluster after the initial configuration. The following request updates the
```

Not all transport settings in the linked doc page can be set for a remote cluster. For example, I don't think you can set `transport.host` or `transport.publish_port` for a remote cluster. So it feels wrong to just link the whole doc page. At the same time, most connection related transport settings should be applicable. If we do want the link, we should also add a brief statement about the limitation.

[[remote-clusters-connect]]
=== Connect to remote clusters
To connect a remote cluster (Cluster A) to a local cluster (Cluster B), you

> configure Cluster A as a remote on Cluster B.

Author: I intentionally duplicated the instructions here so that users can connect a remote cluster without having to link to the CCR page. This way, the instructions apply to both CCS and CCR for connecting a remote cluster.

image::images/connect-remote-clusters.png[ClusterA contains the leader index and ClusterB contains the follower index]

To configure a remote cluster from Stack Management in {kib}:

. Select *Remote Clusters* from the side navigation.
. Specify the {es} endpoint URL, or the IP address or host name of the remote
cluster (`ClusterA`) followed by the transport port (defaults to `9300`). For
example, `cluster.es.eastus2.staging.azure.foundit.no:9400` or
`192.168.1.1:9300`.

Alternatively, use the <<cluster-update-settings,cluster update settings API>>
to add a remote cluster. You can also use this API to 
<<configure-remote-clusters-dynamic,dynamically configure>> remote settings on
every node in the cluster:

[source,console]
----
PUT /_cluster/settings
{
  "persistent" : {
    "cluster" : {
      "remote" : {

>         "leader" : {

Author: I'm clarifying and expanding the lead-in to this example by saying:

> The following request adds a remote cluster with an alias of `leader`. This _cluster alias_ represents the connection to the remote cluster and is used  subsequently to distinguish between local and remote indices. When adding a remote cluster in {kib}, the cluster alias serves as a unique identifier for the remote cluster.

>         "leader" : {

Reviewer: I wonder whether we should avoid using `leader` for the remote cluster. The word `leader` has a specific meaning in CCR for the leader indices. While it kinda makes sense to also call the cluster `leader` for a CCR setup, it does not quite fit for a CCS setup because there is no leader indices for CCS.

>         "leader" : {

Author: Agreed -- I had copied this example over from the CCR docs, but it should be updated to match the surrounding text. I'll change `leader` to `cluster_one` so that it matches the other examples in this section.

You can configure remote clusters settings
<<configure-remote-clusters-dynamic,globally>>, or configure
settings <<configure-remote-clusters-static,on individual nodes>> in the
`elasticsearch.yml` file.

[[configure-remote-clusters-dynamic]]
==== Dynamically configure remote clusters
Use the <<cluster-update-settings,cluster update settings API>> to dynamically
configure remote settings on every node in the cluster. The following request
adds three remote clusters: `cluster_one`, `cluster_two`, and `cluster_three`:

[source,console]
----
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "cluster_one": {
          "seeds": [
            "127.0.0.1:9300"
          ]
        },
        "cluster_two": {
          "mode": "sniff",
          "seeds": [
            "127.0.0.1:9301"
          ],
          "transport.compress": true,
          "skip_unavailable": true
        },
        "cluster_three": {
          "mode": "proxy",
          "proxy_address": "127.0.0.1:9302"
        }
      }
    }
  }
}
----
// TEST[setup:host]
// TEST[s/127.0.0.1:9300/\${transport_host}/]

You can dynamically update the compression and ping schedule settings. However,

> you must include the `seeds` or `proxy_address` in the settings update request.

Author: You seem to be correct. I need to connect remote clusters to effectively test this, but the following request seems to work locally:
```
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "cluster_two": {
          "transport.compress": false
        },
        "cluster_three": {
          "transport.compress": true,
          "transport.ping_schedule": "60s"
        }
      }
    }
  }
}
```

> you must include the `seeds` or `proxy_address` in the settings update request.

Author: >However, you must include the `seeds` or `proxy_address` in the settings update request.

I'm removing this statement, as I agree that it seems incorrect based on my testing.

For example:

[source,console]
----
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "cluster_one": {
          "seeds": [
            "127.0.0.1:9300"
          ]

>         },

Author: I'll remove all parts of the example that are not specific to the updates being made in the request.

          "transport.ping_schedule": "60s"
        }
      }
    }
  }
}
----
// TEST[continued]

NOTE: When the compression or ping schedule settings change, all the existing
node connections must close and re-open, which can cause in-flight requests to
fail.

You can delete a remote cluster from the cluster settings by passing `null`
values for each remote cluster setting. The following request removes
`cluster_two` from the cluster settings, leaving `cluster_one` and 
`cluster_three` intact.:

[source,console]
----
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "cluster_two": {
          "mode": null,
          "seeds": null,
          "skip_unavailable": null,
          "transport.compress": null
        }
      }
    }
  }
}
----
// TEST[continued]

[[configure-remote-clusters-static]]
==== Statically configure remote clusters
If you specify settings in `elasticsearch.yml`, only the nodes with
those settings can connect to the remote cluster and serve remote cluster 
requests.

In the following example, `cluster_one`, `cluster_two`, and `cluster_three` are 
arbitrary _cluster aliases_ representing the connection to each cluster. These 

> names are subsequently used to distinguish between local and remote indices.

Author: Agreed! See my [other comment](https://github.com/elastic/elasticsearch/pull/77043#discussion_r708515942) about this section.

Use the `seeds` parameter to specify the hostname and
<<transport-settings,transport port>> (default `9300`) of a seed node in the 
remote cluster.

The `mode` parameter determines the configured connection mode, which defaults

> to <<sniff-mode,`sniff`>>. Because `cluster_one` doesn't specify a `mode`, it

Author: +1 -- I'll move this explanation to just before the first example of the section to **Dynamically configure remote clusters**.
